File: java/366-find-leaves-of-binary-tree.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
class Solution {
    int low = Integer.MAX_VALUE;
    int high = Integer.MIN_VALUE;
    public List<List<Integer>> findLeaves(TreeNode root) {
        List<List<Integer>> ans = new ArrayList<>();
        if (root == null) {
            return ans;
        }

        getHeight(root, ans);

        return ans;
    }
    private int getHeight(TreeNode node, List<List<Integer>> list) {
        if (node == null) {
            return -1;
        }
        int left = getHeight(node.left, list);
        int right = getHeight(node.right, list);
        int height = Math.max(left, right) + 1;
        if (height >= list.size()) {
            list.add(new ArrayList<>());
        }
        list.get(height).add(node.val);
        return height;
    }
}

File: python3/366.find-leaves-of-binary-tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def findLeaves(self, root: TreeNode) -> List[List[int]]:
        height_to_nodes = collections.defaultdict(list)
        
        def get_height(curr):
            if curr is None:
                return 0
            left = get_height(curr.left)
            right = get_height(curr.right)
            
           
            height = max(left, right) + 1
            height_to_nodes[height].append(curr.val)
            
            return height
        
        get_height(root)
        ans = []
        for k in sorted(height_to_nodes.keys()):
            ans.append(height_to_nodes[k])
            
        return ans
                
                
        
